torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -64,7 +64,9 @@ Reducer::Reducer(
 
   // Initialize variable bucketing.
   // This can be reinitialized later after capturing runtime information.
+  std::unique_lock<std::mutex> lock(this->mutex_);
```

Comment:
lock was inside initializeBuckets() before, I am moving this lock out of it here. So that we do not need to unlock before calling intializeBuckets() inside rebuildBuckets() (the unlock may result in race with other threads)

Reply:
hmm, which thread might lead to race here? I don't clearly recall the original reason for adding lock to `initialize_buckets`, was it because it was used in `mark_variable_ready`? After we move it to Reducer ctor here, is this lock still necessary?

Reply:
initialize_buckets has lock originally even when it is called in reducer constructor only. I think the lock is added possibly because initialize_buckets was exposed to be reducer's public function, so it could be called by multiple threads and thus lock is needed in this public function. 

I agree with you that the lock may be redundant in constructor. But I'm still keeping the lock here to keep the same logic as before, just in case I missed something